README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->

[![R-CMD-check](https://github.com/ateucher/lutz/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/ateucher/lutz/actions/workflows/R-CMD-check.yaml)
[![Coverage
Status](https://img.shields.io/codecov/c/github/ateucher/lutz/master.svg)](https://app.codecov.io/github/ateucher/lutz?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/lutz)](https://cran.r-project.org/package=lutz)
[![CRAN
downloads](https://cranlogs.r-pkg.org/badges/lutz)](https://cran.r-project.org/package=lutz)
<!-- badges: end -->

# lutz (look up time zones)

## Lookup the time zone of coordinates

Input latitude and longitude values or an `sf/sfc` POINT object and get
back the time zone in which they exist. Two methods are implemented. One
is very fast and uses Rcpp in conjunction with source data from
(<https://github.com/darkskyapp/tz-lookup-oss/>). However, speed comes
at the cost of accuracy - near time zone borders away from populated
centres there is a chance that it will return the incorrect time zone.

The other method is slower but more accurate - it uses the sf package to
intersect points with a detailed map of time zones from
[here](https://github.com/evansiroky/timezone-boundary-builder).

## time zone utility functions

**lutz** also contains several utility functions for helping to
understand and visualize time zones, such as listing of world time
zones,including information about daylight savings times and their
offsets from UTC. You can also plot a time zone to visualize the UTC
offset over a year and when daylight savings times are in effect.

## Installation

You can install lutz from CRAN with:

``` r
install.packages("lutz")
```

Or you can install the development version from github with:

``` r
# install.packages("devtools")
devtools::install_github("ateucher/lutz")
```

## Examples

There are two functions in this package for looking up the time zones of
coordinates: `tz_lookup()` which works with both `sf/sfc` and
`SpatialPoints` objects, and `tz_lookup_coords` for looking up lat/long
pairs. Use the `method` argument to choose the `"fast"` or `"accurate"`
method.

### With coordinates. They must be lat/long in decimal degrees:

``` r
library(lutz)
tz_lookup_coords(49.5, -123.5, method = "fast")
#> [1] "America/Vancouver"
tz_lookup_coords(49.5, -123.5, method = "accurate")
#> [1] "America/Vancouver"

tz_lookup_coords(lat = c(48.9, 38.5, 63.1, -25), lon = c(-123.5, -110.2, -95.0, 130))
#> [1] "America/Vancouver"    "America/Denver"       "America/Rankin_Inlet"
#> [4] "Australia/Darwin"
```

### With `sf` objects:

``` r
library(sf)
library(ggplot2) # this requires the devlopment version of ggplot2

# Create an sf object out of the included state.center dataset:
pts <- lapply(seq_along(state.center$x), function(i) {
  st_point(c(state.center$x[i], state.center$y[i]))
})
state_centers_sf <- st_sf(st_sfc(pts))

# Use tz_lookup_sf to find the time zones
state_centers_sf$tz <- tz_lookup(state_centers_sf)
state_centers_sf$tz <- tz_lookup(state_centers_sf, method = "accurate")

ggplot() + 
  geom_sf(data = state_centers_sf, aes(colour = tz)) + 
  theme_minimal() + 
  coord_sf(datum = NA)
```

![](man/figures/unnamed-chunk-4-1.png)<!-- -->

### With `SpatialPoints` objects:

``` r
library(sp)
state_centers_sp <- as(state_centers_sf, "Spatial")

state_centers_sp$tz <- tz_lookup(state_centers_sp)

ggplot(cbind(as.data.frame(coordinates(state_centers_sp)), tz = state_centers_sp$tz), 
       aes(x = coords.x1, y = coords.x2, colour = tz)) + 
  geom_point() + 
  coord_fixed() + 
  theme_minimal()
```

![](man/figures/unnamed-chunk-5-1.png)<!-- -->

Note that there are some regions in the world where a single point can
land in two different overlapping time zones. The `"accurate"` method
[includes
these](https://github.com/evansiroky/timezone-boundary-builder/releases/tag/2018g),
however the method used in the `"fast"` does not include overlapping
time zones ([at least for
now](https://github.com/darkskyapp/tz-lookup-oss/issues/34)).

We can compare the accuracy of both methods to the high-resolution time
zone map provided by
<https://github.com/evansiroky/timezone-boundary-builder>. This is the
map that is used by `lutz` for the `"accurate"` method, but in `lutz` it
is simplified by about 80% to be small enough to fit in the package.

``` r
## Get the full time zone geojson from https://github.com/evansiroky/timezone-boundary-builder
download.file("https://github.com/evansiroky/timezone-boundary-builder/releases/download/2019a/timezones-with-oceans.geojson.zip",
                destfile = "tz.zip")
unzip("tz.zip", exdir = "data-raw/dist/")
```

``` r
library(lutz)
library(sf)
library(purrr)
library(dplyr)

tz_full <- read_sf("data-raw/dist/combined-with-oceans.json")
# Create a data frame of 500000 lat/long pairs:
set.seed(1)
n <- 500000
ll <- data.frame(id = seq(n), lat = runif(n, -90, 90), lon = runif(n, -180, 180))
ll_sf <- st_as_sf(ll, coords = c("lon", "lat"), crs = 4326)

# Overlay those points with the full high-resolution time zone map:
ref_ll_tz <- sf::st_join(ll_sf, tz_full)

# Combine those that had overlapping time zones
ref_ll_tz <- ref_ll_tz %>% 
  st_set_geometry(NULL) %>% 
  group_by(id) %>% 
  summarize(tzid = paste(tzid, collapse = "; "))

# run tz_lookup with both `"fast"` and `"accurate"` methods and compare with 
# the time zones looked up with the high-resolution map:
tests <- map_df(c("fast", "accurate"), ~ {
  time <- system.time(test_ll_tz <- tz_lookup(ll_sf, method = .x, warn = FALSE))
  comp <- ref_ll_tz$tzid == test_ll_tz
  matches <- sum(comp, na.rm = TRUE)
  mismatches <- sum(!comp, na.rm = TRUE)
  list(
    method = .x,
    time = time["elapsed"],
    matches = matches,
    mismatches = mismatches,
    accuracy = matches / (matches + mismatches),
    ref_nas = sum(is.na(ref_ll_tz$tzid)),
    fun_nas = sum(is.na(test_ll_tz))
    )
})
```

``` r
knitr::kable(tests)
```

## time zone utility functions

### `tz_plot()`

``` r
tz_plot("America/Vancouver")
```

![](man/figures/unnamed-chunk-9-1.png)<!-- -->

### `tz_offset()`

``` r
# A Date object
tz_offset(Sys.Date(), "Africa/Algiers")
#>          tz_name  date_time zone is_dst utc_offset_h
#> 1 Africa/Algiers 2023-10-17  CET  FALSE            1


# A Date-like character string
tz_offset("2017-03-01", tz = "Singapore")
#>     tz_name  date_time zone is_dst utc_offset_h
#> 1 Singapore 2017-03-01  +08  FALSE            8


# A POSIXct date-time object
tz_offset(Sys.time())
#> Warning: You supplied an object of class POSIXct that does not have a time zone
#> attribute, and did not specify one inthe 'tz' argument. Defaulting to current
#> (America/Vancouver).
#>             tz_name           date_time zone is_dst utc_offset_h
#> 1 America/Vancouver 2023-10-17 12:11:11  PDT   TRUE           -7
```

### `tz_list()`

``` r
tz_list() %>% 
  head(20) %>% 
  knitr::kable()
```

|     | tz_name            | zone | is_dst | utc_offset_h |
|:----|:-------------------|:-----|:-------|-------------:|
| 1   | Africa/Abidjan     | GMT  | FALSE  |            0 |
| 2   | Africa/Accra       | GMT  | FALSE  |            0 |
| 3   | Africa/Addis_Ababa | EAT  | FALSE  |            3 |
| 4   | Africa/Algiers     | CET  | FALSE  |            1 |
| 5   | Africa/Asmara      | EAT  | FALSE  |            3 |
| 6   | Africa/Asmera      | EAT  | FALSE  |            3 |
| 7   | Africa/Bamako      | GMT  | FALSE  |            0 |
| 8   | Africa/Bangui      | WAT  | FALSE  |            1 |
| 9   | Africa/Banjul      | GMT  | FALSE  |            0 |
| 10  | Africa/Bissau      | GMT  | FALSE  |            0 |
| 11  | Africa/Blantyre    | CAT  | FALSE  |            2 |
| 12  | Africa/Brazzaville | WAT  | FALSE  |            1 |
| 13  | Africa/Bujumbura   | CAT  | FALSE  |            2 |
| 14  | Africa/Cairo       | EET  | FALSE  |            2 |
| 16  | Africa/Cairo       | EEST | TRUE   |            3 |
| 17  | Africa/Casablanca  | +01  | TRUE   |            1 |
| 18  | Africa/Casablanca  | +00  | FALSE  |            0 |
| 19  | Africa/Ceuta       | CET  | FALSE  |            1 |
| 20  | Africa/Ceuta       | CEST | TRUE   |            2 |
| 21  | Africa/Conakry     | GMT  | FALSE  |            0 |


NEWS.md
# lutz (development version)

# lutz 0.3.2

This is a maintenance release, with no new features. The rgdal package was 
removed as a Suggested package, as it has been archived on CRAN.

# lutz 0.3.1

* Fixed bug where invalid timezones would cause `tz_list()` to fail.

# lutz 0.3.0

* V8-based 'fast' method was ported to use Rcpp - it is now even faster, and
  we can drop the V8 dependency. Thanks @hrbrmstr! (#4)
* Bob Rudis (@hrbrmstr) added as an author
* Upgraded time zone map to 2019b
* Dealt with areas with overlapping time zones (#2)
* Added three new functions:
  - tz_list() lists all time zones and information about UTC offsets and daylight savings
  - tz_offset() lists information about a specific time in a specific time zone
  - tz_plot() plots a time zone and its UTC offset for a year, including periods
  of daylight savings.

# lutz 0.2.0

* Added `method = "accurate"` to do a slower, but more accurate lookup.
* Updated timezone map

# lutz 0.1.0

* Initial release


R/tz_lookup.R
#' Lookup time zones of sf or sp points
#'
#' There are two methods - `"fast"`, and `"accurate"`. The `"fast"` version can
#' look up many thousands of points very quickly, however  when a point is near
#' a time zone boundary and not near a populated centre, it may return the
#' incorrect time zone. If accuracy is more important than speed, use
#' `method = "accurate"`.
#'
#' Note that there are some regions in the world where a single point can land in
#' two different overlapping time zones. The `"accurate"` method includes these,
#' and when they are encountered they are concatenated in a single string,
#' separated by a semicolon.
#' The data used in the `"fast"` method does not include overlapping time zones
#' at this time.
#'
#' @param x either an `sfc` or `sf` points or `SpatialPoints(DataFrame)` object
#' @param crs the coordinate reference system: integer with the EPSG code, or character with proj4string.
#' If not specified (i.e., `NULL`) and `x` has no existing `crs`, EPSG: 4326 is assumed (lat/long).
#' @param method method by which to do the lookup. Either `"fast"` (default)
#' or `"accurate"`.
#' @param warn By default, if `method = "fast"` a warning is issued about
#'   the potential for inaccurate results. Set `warn` to `FALSE` to turn
#'   this off.
#'
#' @return character vector the same length as `x` specifying the time zone of the points.
#' @export
#'
#' @examples
#' if (require("sf")) {
#'
#' state_pts <- lapply(seq_along(state.center$x), function(i) {
#'   st_point(c(state.center$x[i], state.center$y[i]))
#' })
#'
#' state_centers_sf <- st_sf(st_sfc(state_pts))
#'
#' state_centers_sf$tz <- tz_lookup(state_centers_sf)
#'
#' plot(state_centers_sf[, "tz"])
#' }
#'
tz_lookup <- function(x, crs = NULL, method = "fast", warn = TRUE) {
  switch(method,
         fast = tz_lookup_fast(x, crs, warn),
         accurate = tz_lookup_accurate(x, crs),
         stop("method must be one of 'fast' or 'accurate'", call. = FALSE))

}

tz_lookup_fast <- function(x, crs = NULL, warn) {
  UseMethod("tz_lookup_fast")
}

#' @importFrom sf st_coordinates
tz_lookup_fast.sf <- function(x, crs, warn) {
  x <- fix_sf(x, crs)
  coords <- sf::st_coordinates(x)
  # sf stores as x, y and tzlookup likes lat, lon (Which is the opposite)
  tz_lookup_coords_fast(lat = coords[, 2], lon = coords[, 1], warn = warn)
}

tz_lookup_fast.sfc <- tz_lookup_fast.sf

#' @importFrom sp coordinates
tz_lookup_fast.SpatialPoints <- function(x, crs, warn) {
  x <- fix_sp(x, crs)

  coords <- sp::coordinates(x)
  # sf stores as x, y and tzlookup likes lat, lon (Which is the opposite)
  tz_lookup_coords_fast(lat = coords[, 2], lon = coords[, 1], warn = warn)
}

#' Lookup time zones of lat/long pairs
#'
#' There are two methods - `"fast"`, and `"accurate"`. The `"fast"` version can
#' look up many thousands of points very quickly, however  when a point is near
#' a time zone boundary and not near a populated centre, it may return the
#' incorrect time zone. If accuracy is more important than speed, use
#' `method = "accurate"`.
#'
#' @param lat numeric vector of latitudes
#' @param lon numeric vector of longitudes the same length as `x`
#' @inheritParams tz_lookup
#' @return character vector the same length as x and y specifying the time zone of the points.
#' @export
#'
#' @examples
#' tz_lookup_coords(42, -123)
#' tz_lookup_coords(lat = c(48.9, 38.5, 63.1, -25), lon = c(-123.5, -110.2, -95.0, 130))
tz_lookup_coords <- function(lat, lon, method = "fast", warn = TRUE) {
  check_for_spatial(lat)
  check_coords(lat, lon)

  switch(method,
         fast = tz_lookup_coords_fast(lat, lon, warn),
         accurate = tz_lookup_coords_accurate(lat, lon),
         stop("method must be one of 'fast' or 'accurate'", call. = FALSE)
  )
}

tz_lookup_coords_fast <- function(lat, lon, warn) {

  if (warn) warn_for_fast()

  timezone_lookup_coords_rcpp(lat, lon)

}

tz_lookup_accurate <- function(x, crs = NULL) {
  UseMethod("tz_lookup_accurate")
}

#' @importFrom sf st_set_geometry st_join
#' @importFrom stats aggregate
tz_lookup_accurate.sf <- function(x, crs = NULL) {
  x <- fix_sf(x, crs)
  # Add a unique id so we can deal with any duplicates resulting
  # from overlapping time zones
  x$lutzid <- seq_len(nrow(x))
  x_tz <- suppressMessages(sf::st_set_geometry(sf::st_join(x, tz_sf), NULL))

  # group x by lutzid and concatenate multiple time zones with ;
  if (nrow(x_tz) > nrow(x)) {
    warning("Some points are in areas with more than one time zone defined.",
            "These are often disputed areas and should be treated with care.")

    ret <- stats::aggregate(x_tz, list(x_tz$lutzid), function(x) {
      if (length(x) == 1) return(x)
      x <- paste(x, collapse = "; ")
    }, drop = FALSE)[["tzid"]]

  } else {
    ret <- x_tz$tzid
  }

  # If any are NA, try to fill in with Rcpp-based tzlookup
  nas <- which(is.na(ret))
  if (!length(nas)) {
    return(ret)
  }
  ret[nas] <- tz_lookup_fast(x[nas, ], warn = FALSE) # nocov start
  ret # nocov end
}

#' @importFrom sf st_sf
tz_lookup_accurate.sfc <- function(x, crs = NULL) {
  x_sf <- sf::st_sf(geom = x)
  tz_lookup_accurate(x_sf, crs = crs)
}

#' @importFrom sf st_as_sf
tz_lookup_accurate.SpatialPoints <- function(x, crs = NULL) {
  x <- sf::st_as_sf(x)
  tz_lookup_accurate(x, crs)
}

#' @importFrom sf st_as_sf
#' @importFrom stats complete.cases
tz_lookup_coords_accurate <- function(lat, lon) {

  ll <- data.frame(lat = lat, lon = lon)
  # check for NAs before converting to sf, only send valid lat/lon pairs
  # into sf, recombine with NAs for output
  cc <- stats::complete.cases(ll)
  ret <- rep(NA_character_, length(cc))
  if (sum(cc)) {
    ll_sf <- sf::st_as_sf(ll[cc, ], coords = c("lon", "lat"), crs = 4326)
    ret[cc] <- tz_lookup_accurate(ll_sf)
  }
  ret
}
